vms/platformvm/state/validator_metadata.go

```diff
@@ -0,0 +1,265 @@
+	// [subnetID].
+	GetDelegateeReward(
+		subnetID ids.ID,
+		vdrID ids.NodeID,
+	) (amount uint64, err error)
+
+	// SetDelegateeReward updates the rewards accrued to [vdrID] on [subnetID].
+	// Unless these measurements are deleted first, the next call to
+	// WriteUptimes will write this update to disk.
+	SetDelegateeReward(
+		subnetID ids.ID,
+		vdrID ids.NodeID,
+		amount uint64,
+	) error
+
+	// DeleteValidatorMetadata removes in-memory references to the metadata of
+	// [vdrID] on [subnetID]. If there were staged updates from a prior call to
+	// SetUptime or SetDelegateeReward, the updates will be dropped. This call
+	// will not result in a write to disk.
+	DeleteValidatorMetadata(vdrID ids.NodeID, subnetID ids.ID)
+
+	// WriteValidatorMetadata writes all staged updates from prior calls to
+	// SetUptime or SetDelegateeReward.
+	WriteValidatorMetadata(
+		dbPrimary database.KeyValueWriter,
+		dbSubnet database.KeyValueWriter,
+	) error
+}
+
+type metadata struct {
+	metadata map[ids.NodeID]map[ids.ID]*validatorMetadata // vdrID -> subnetID -> metadata
+	// updatedMetadata tracks the updates since WriteValidatorMetadata was last called
+	updatedMetadata map[ids.NodeID]set.Set[ids.ID] // vdrID -> subnetIDs
+}
+
+func newValidatorState() validatorState {
+	return &metadata{
+		metadata:        make(map[ids.NodeID]map[ids.ID]*validatorMetadata),
+		updatedMetadata: make(map[ids.NodeID]set.Set[ids.ID]),
+	}
+}
+
+func (m *metadata) LoadValidatorMetadata(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+	uptime *validatorMetadata,
+) {
+	subnetMetadata, ok := m.metadata[vdrID]
+	if !ok {
+		subnetMetadata = make(map[ids.ID]*validatorMetadata)
+		m.metadata[vdrID] = subnetMetadata
+	}
+	subnetMetadata[subnetID] = uptime
+}
+
+func (m *metadata) GetUptime(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+) (time.Duration, time.Time, error) {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return 0, time.Time{}, database.ErrNotFound
+	}
+	return metadata.UpDuration, metadata.lastUpdated, nil
+}
+
+func (m *metadata) SetUptime(
+	vdrID ids.NodeID,
+	subnetID ids.ID,
+	upDuration time.Duration,
+	lastUpdated time.Time,
+) error {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return database.ErrNotFound
+	}
+	metadata.UpDuration = upDuration
+	metadata.lastUpdated = lastUpdated
+
+	m.addUpdatedMetadata(vdrID, subnetID)
+	return nil
+}
+
+func (m *metadata) GetDelegateeReward(
+	subnetID ids.ID,
+	vdrID ids.NodeID,
+) (uint64, error) {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return 0, database.ErrNotFound
+	}
+	return metadata.PotentialDelegateeReward, nil
+}
+
+func (m *metadata) SetDelegateeReward(
+	subnetID ids.ID,
+	vdrID ids.NodeID,
+	amount uint64,
+) error {
+	metadata, exists := m.metadata[vdrID][subnetID]
+	if !exists {
+		return database.ErrNotFound
+	}
+	metadata.PotentialDelegateeReward = amount
+
+	m.addUpdatedMetadata(vdrID, subnetID)
+	return nil
+}
```

Comment:
I know these are really similar to `Get/SetUptime` but would be good to add a couple tests for this functionality here as well.